includes/class-llms-order-generator.php

```diff
@@ -0,0 +1,733 @@
+			if ( is_wp_error( $res ) ) {
+				return $res;
+			}
+		}
+
+		/**
+		 * Allows 3rd party prior to generation of an order.
+		 *
+		 * This validation hooks runs after to all default validation passes.
+		 *
+		 * @since [version]
+		 *
+		 * @param boolean|WP_Error $validation_errors Halts checkout and returns the supplied error.
+		 */
+		return apply_filters( 'llms_after_generate_order_validation', true );
+
+	}
+
+	/**
+	 * Validates the coupon.
+	 *
+	 * @since [version]
+	 *
+	 * @return boolean|WP_Error Returns `true` on success or an error object.
+	 */
+	protected function validate_coupon() {
+
+		// If a coupon is being used, validate it.
+		if ( ! empty( $this->data['llms_coupon_code'] ) ) {
+
+			$code = sanitize_text_field( $this->data['llms_coupon_code'] );
+
+			// Locate the coupon post ID.
+			$coupon_id = llms_find_coupon( $code );
+			if ( ! $coupon_id ) {
+				return $this->error(
+					self::E_COUPON_NOT_FOUND,
+					sprintf(
+						// Translators: %s = The user-submitted coupon code.
+						__( 'Coupon code "%s" not found.', 'lifterlms' ),
+						$code
+					)
+				);
+			}
+
+			// Validate the coupon for the current plan.
+			$coupon = llms_get_post( $coupon_id );
+			$valid  = $coupon->is_valid( $this->plan->get( 'id' ) );
+			if ( is_wp_error( $valid ) ) {
+				return $this->error( self::E_COUPON_INVALID, $valid->get_error_message() );
+			}
+
+			$this->coupon = $coupon;
+
+		}
+
+		return true;
+	}
+
+	/**
+	 * Validates the payment gateway.
+	 *
+	 * @since [version]
+	 *
+	 * @return boolean|WP_Error Returns `true` on success or an error object.
+	 */
+	protected function validate_gateway() {
+
+		$coupon_id = $this->coupon ? $this->coupon->get( 'id' ) : null;
+
+		/**
+		 * If payment is required, verify we have a gateway.
+		 *
+		 * For free plans the manual gateway is automatically used, whether or not it's enabled.
+		 */
+		if ( $this->plan->requires_payment( $coupon_id ) && empty( $this->data['llms_payment_gateway'] ) ) {
+			return $this->error( self::E_GATEWAY_REQUIRED, __( 'No payment method selected.', 'lifterlms' ) );
+		}
+
+		$gateway_id = $this->data['llms_payment_gateway'] ?? 'manual';
+		$is_valid   = llms_can_gateway_be_used_for_plan( $gateway_id, $this->plan );
+		if ( is_wp_error( $is_valid ) ) {
+			return $is_valid;
+		}
+
+		$this->gateway = llms()->payment_gateways()->get_gateway_by_id( $gateway_id );
+		return true;
+
+	}
+
+	/**
+	 * Validates the order.
+	 *
+	 * Ensures the submitted order key is valid and that the order can be confirmed.
+	 *
+	 * @since [version]
+	 *
+	 * @return boolean|WP_Error Returns `true` on success or an error object.
+	 */
+	protected function validate_order() {
+
+		$order_id = $this->get_order_id();
+
+		if ( 'new' === $order_id || 'llms_order' !== get_post_type( $order_id ) ) {
+			return $this->error(
+				self::E_ORDER_NOT_FOUND,
+				__( 'Could not locate an order to confirm.', 'lifterlms' )
+			);
+		}
+
+		$order = llms_get_post( $order_id );
```

Comment:
Might want to do a sanity check in case `llms_get_post()` returns `null` or `false`.

Reply:
`get_order_id()` runs that sanity check for us